### PR TITLE
Introduce an identifier for the default source/target for a collection

### PR DESCRIPTION
**Pull request by isuruf**
And rename DEFAULT_SOURCE -> UNNAMED_SOURCE

### CODE REVIEW

**Comment by alexfikl** [comment]
This looks good to me, but could use a bit more documentation. Some things that come to mind:
* Why can't we use `None` for `UNNAMED_SOURCE` and `UNNAMED_TARGET`?
* `DEFAULT_SOURCE` and `DEFAULT_TARGET` are replaced in `LocationTagger` with whatever it was given in the constructor, so not necessarily the `auto_source` and `auto_target` of the `GeometryCollection`. This disagrees with their docs?
* Are `DEFAULT_SOURCE` and `DEFAULT_TARGET` supposed to be for internal use now? i.e. stuff in `pytential.symbolic`.

**Comment by isuruf** [comment]
> Why can't we use None for UNNAMED_SOURCE and UNNAMED_TARGET?

They need to be distinct values in order to be placed in a dictionary.

> DEFAULT_SOURCE and DEFAULT_TARGET are replaced in LocationTagger with whatever it was given in the constructor, so not necessarily the auto_source and auto_target of the GeometryCollection. This disagrees with their docs?

Fixed now.

> Are DEFAULT_SOURCE and DEFAULT_TARGET supposed to be for internal use now? i.e. stuff in pytential.symbolic.

Nope. UNNAMED_SOURCE and UNNAMED_TARGET are internal use only. Added to docs.

File: pytential/collection.py [hunk 1]
```diff
@@ -142,8 +142,8 @@ def __init__(self,
             descriptors are used to define the default source and target
             geometries for layer potential evaluations.
             By default, they are set to
-            :class:`~pytential.symbolic.dof_desc.DEFAULT_SOURCE` and
-            :class:`~pytential.symbolic.dof_desc.DEFAULT_TARGET` for
+            :class:`~pytential.symbolic.dof_desc.UNNAMED_SOURCE` and
+            :class:`~pytential.symbolic.dof_desc.UNNAMED_TARGET` for
```
Comment (inducer) [inline pytential/collection.py:146]: ...unspecified values...

File: pytential/symbolic/dof_desc.py [hunk 2]
```diff
@@ -23,6 +23,9 @@
 from typing import Any, Hashable, Optional, Union
 
 __doc__ = """
+.. autoclass:: UNNAMED_SOURCE
+.. autoclass:: UNNAMED_TARGET
+
```
Comment (inducer) [inline pytential/symbolic/dof_desc.py:28]:
```suggestion
```

File: pytential/symbolic/dof_desc.py [hunk 3]
```diff
@@ -55,12 +58,26 @@
 
 # {{{ discretizations
 
+class UNNAMED_SOURCE:                   # noqa: N801
```
Comment (inducer) [inline pytential/symbolic/dof_desc.py:61]: Link to the use case in the interface for `prepare_...`. (same for `_UNNAMED_TARGET`)

**Review by inducer — APPROVED** [review]
LGTM subject to the changes above.

File: pytential/symbolic/execution.py [hunk 4]
```diff
@@ -37,6 +37,7 @@
 from pytential.symbolic.compiler import Code, Statement, Assign, ComputePotential
 
 from pytential import sym
+from pytential.symbolic.dof_desc import UNNAMED_SOURCE, UNNAMED_TARGET
 
 import logging
```
Comment (inducer) [inline pytential/symbolic/execution.py:42]: - Delete `domains` arg on `build_matrix`
- Delete `domains` arg on `scipy_op`.

Comment (alexfikl) [reply pytential/symbolic/execution.py:42]: Why? Weren't those supposed to support multiple geometries with different properties? (which, yeah, nobody uses or tests at all.. but still!)

Comment (inducer) [reply pytential/symbolic/execution.py:42]: Hmmm, upon closer reflection, I think I may have jumped the gun on removing `domains`. I think it has a role to play still. What tripped me up is that there should be `source_domains` and `target_domains`, technically. Where it says `DEFAULT_TARGET` must be a key, instead say something like "target" must be the result of an unnamed binding, such as might result from passing a tuple to `bind`.

IOW, @alexfikl, you're completely right.

**Review by alexfikl — APPROVED** [review]

**Comment by inducer** [comment]
Thx!

**Pull request merged**